File: rust-version/src/camera.rs
use crate::hittable::{HitRecord, Hittable};
use crate::raytracing::*;

pub struct Camera {
    image_width: usize,
    image_height: usize,
    samples_per_pixel: usize,
    max_depth: usize,
    center: Point3<f64>,
    pixel00_loc: Point3<f64>,
    pixel_delta_u: Vector3<f64>,
    pixel_delta_v: Vector3<f64>,
    pixel_samples_scale: f64,
    defocus_angle: f64,
    defocus_disk_u: Vector3<f64>,
    defocus_disk_v: Vector3<f64>,
}

impl Camera {
    pub fn new(
        aspect_ratio: f64,
        image_width: usize,
        samples_per_pixel: usize,
        max_depth: usize,
        vfov: f64,
        lookfrom: Point3<f64>,
        lookat: Point3<f64>,
        vup: Vector3<f64>,
        defocus_angle: f64,
        focus_dist: f64,
    ) -> Self {
        let image_height = (image_width as f64 / aspect_ratio) as usize;
        let image_height = image_height;

        let pixel_samples_scale = 1.0 / samples_per_pixel as f64;

        let center = lookfrom;

        // Determine viewport dimensions.
        let theta = degrees_to_radians(vfov);
        let h = (theta / 2.0).tan();
        let viewport_height = 2.0 * h * focus_dist;
        let viewport_width = viewport_height * (image_width as f64 / image_height as f64);

        // Calculate the u,v,w unit basis vectors for the camera coordinate frame.
        let w = unit_vector(lookfrom - lookat);
        let u = unit_vector(vup.cross(&w));
        let v = w.cross(&u);

        // Calculate the vectors across the horizontal and down the vertical viewport edges.
        let viewport_u = viewport_width * u;
        let viewport_v = viewport_height * -v;

        // Calculate the horizontal and vertical delta vectors from pixel to pixel.
        let pixel_delta_u = viewport_u / image_width as f64;
        let pixel_delta_v = viewport_v / image_height as f64;

        // Calculate the location of the upper left pixel.
        let viewport_upper_left = center - (focus_dist * w) - viewport_u / 2.0 - viewport_v / 2.0;
        let pixel00_loc = viewport_upper_left + 0.5 * (pixel_delta_u + pixel_delta_v);

        // Calculate the camera defocus disk basis vectors.
        let defocus_radius = focus_dist * (degrees_to_radians(defocus_angle / 2.0)).tan();
        let defocus_disk_u = u * defocus_radius;
        let defocus_disk_v = v * defocus_radius;

        Self {
            image_width,
            image_height,
            samples_per_pixel,
            max_depth,
            center,
            pixel00_loc,
            pixel_delta_u,
            pixel_delta_v,
            pixel_samples_scale,
            defocus_angle,
            defocus_disk_u,
            defocus_disk_v,
        }
    }

    pub fn ray_color(&self, r: &Ray, depth: usize, world: &dyn Hittable) -> Color {
        if depth <= 0 {
            return Color::new(0.0, 0.0, 0.0);
        }

        let mut rec = HitRecord::default();
        if world.hit(r, Interval::new(0.001, INFINITY), &mut rec) {
            let mut scattered = Ray::default();
            let mut attenuation = Color::new(0.0, 0.0, 0.0);
            if let Some(material) = &rec.mat {
                if material.scatter(r, &rec, &mut attenuation, &mut scattered) {
                    return attenuation.component_mul(&self.ray_color(
                        &scattered,
                        depth - 1,
                        world,
                    ));
                }
                return Color::new(0.0, 0.0, 0.0);
            }
        }

        let unit_direction = r.direction().normalize();
        let a = 0.5 * (unit_direction.y + 1.0);
        (1.0 - a) * Color::new(1.0, 1.0, 1.0) + a * Color::new(0.5, 0.7, 1.0)
    }

    fn get_ray(&self, i: usize, j: usize) -> Ray {
        let offset = self.sample_square();
        let pixel_sample = self.pixel00_loc
            + (i as f64 + offset.x) * self.pixel_delta_u
            + (j as f64 + offset.y) * self.pixel_delta_v;

        let ray_origin = if self.defocus_angle <= 0.0 {
            self.center
        } else {
            self.defocus_disk_sample()
        };
        let ray_direction = pixel_sample - ray_origin;

        Ray::new(ray_origin, ray_direction.normalize())
    }

    fn sample_square(&self) -> Vector3<f64> {
        Vector3::new(random_double() - 0.5, random_double() - 0.5, 0.0)
    }

    fn defocus_disk_sample(&self) -> Point3<f64> {
        let p = random_in_unit_disk();
        self.center + (p.x * self.defocus_disk_u) + (p.y * self.defocus_disk_v)
    }

    pub fn render(&self, world: &dyn Hittable) {
        println!("P3\n{} {}\n255", self.image_width, self.image_height);

        for j in 0..self.image_height {
            eprint!("\rScanlines remaining: {} ", self.image_height - j - 1);
            for i in 0..self.image_width {
                let mut pixel_color = Color::new(0.0, 0.0, 0.0);
                for _ in 0..self.samples_per_pixel {
                    let r = self.get_ray(i, j);
                    pixel_color += self.ray_color(&r, self.max_depth, world);
                }
                write_color(
                    &mut std::io::stdout(),
                    self.pixel_samples_scale * pixel_color,
                )
                .unwrap();
            }
        }

        eprint!("\n\rDone.\n");
    }
}

File: Engine/src/Camera/camera.cpp
#include "camera.hpp"

Camera::Camera(double aspect_ratio, int image_width, int samples_per_pixel,
               int max_depth, double vfov, double defocus_angle,
               double focus_dist, const point3 &lookfrom, const point3 &lookat,
               const vec3 &vup)
    : image_width(image_width), samples_per_pixel(samples_per_pixel),
      max_depth(max_depth), vfov(vfov), aspect_ratio(aspect_ratio),
      defocus_angle(defocus_angle), focus_dist(focus_dist), lookfrom(lookfrom),
      lookat(lookat), vup(vup), yaw(-90.0), pitch(0.0), zoom(1),
      perlin_noise(make_shared<Perlin>()) {
    initialize();
}

void Camera::render(
    const BVHNode &world, const HittableList &lights, const color &background,
    int j, std::function<void(int, double, const color &)> draw_pixel) {
    for (int i = 0; i < image_width; ++i) {
        color pixel_color(0, 0, 0);
        for (int s_j = 0; s_j < sqrt_spp; ++s_j) {
            for (int s_i = 0; s_i < sqrt_spp; ++s_i) {
                std::pair<int, int> pixel_coord = {j, i};
                if (pixel_cache.find(pixel_coord) != pixel_cache.end()) {
                    pixel_color += pixel_cache[pixel_coord];
                } else {
                    Ray r = getRay(i, j, s_i, s_j);
                    pixel_color +=
                        rayColor(r, background, max_depth, world, lights);
                    pixel_cache[pixel_coord] = pixel_color;
                }
            }
        }
        draw_pixel(i, pixel_samples_scale, pixel_color);
    }
}

void Camera::renderTile(
    int x0, int y0, int x1, int y1, const BVHNode &world,
    const HittableList &lights, const color &background,
    std::function<void(int, int, int, int, double, const color &)> draw_tile) {
    for (int j = y0; j < y1; ++j) {
        for (int i = x0; i < x1; ++i) {
            color pixel_color(0, 0, 0);
            for (int s_j = 0; s_j < sqrt_spp; ++s_j) {
                for (int s_i = 0; s_i < sqrt_spp; ++s_i) {
                    std::pair<int, int> pixel_coord = {j, i};
                    if (pixel_cache.find(pixel_coord) != pixel_cache.end()) {
                        pixel_color += pixel_cache[pixel_coord];
                    } else {
                        auto r = getRay(i, j, s_i, s_j);
                        color sample_color =
                            rayColor(r, background, max_depth, world, lights);
                        pixel_cache[pixel_coord] = sample_color;
                        pixel_color += sample_color;
                    }
                }
            }
            draw_tile(i, j, i + 1, j + 1, pixel_samples_scale, pixel_color);
        }
    }
}

color Camera::rayColor(const Ray &r, color background, int depth,
                       const HitTable &world, const HitTable &lights) const {
    if (depth <= 0) {
        return color(0, 0, 0);
    }

    HitRecord rec;

    if (!world.hit(r, Interval(0.001, infinity), rec)) {
        return background;
    }

    ScatterRecord srec;
    color color_from_emission =
        rec.mat_ptr->emitted(r, rec, rec.u, rec.v, rec.p);

    if (!rec.mat_ptr->scatter(r, rec, srec)) {
        return color_from_emission;
    }

    if (srec.skip_pdf) {
        return srec.attenuation * rayColor(srec.skip_pdf_ray, background,
                                           depth - 1, world, lights);
    }

    auto light_ptr = make_shared<HittablePDF>(lights, rec.p);
    MixturePDF p(light_ptr, srec.pdf_ptr);

    Ray scattered(rec.p, p.generate(), r.time());
    auto pdf_val = p.value(scattered.direction());

    double scattering_pdf = rec.mat_ptr->scatteringPDF(r, rec, scattered);

    color sample_color =
        rayColor(scattered, background, depth - 1, world, lights);
    color color_from_scatter =
        (srec.attenuation * scattering_pdf * sample_color) / pdf_val;

    return color_from_emission + color_from_scatter;
}

void Camera::updateCameraVectors() {
    vec3 front;
    front.x = std::cos(radians(yaw)) * std::cos(radians(pitch));
    front.y = std::sin(radians(pitch));
    front.z = std::sin(radians(yaw)) * std::cos(radians(pitch));
    front = unitVector(front);

    w = unitVector(lookfrom - (lookfrom + front));
    u = unitVector(cross(vup, w));
    v = cross(w, u);

    auto theta = radians(vfov);
    auto h = std::tan(theta / 2);
    auto viewport_height = 2.0 * h * focus_dist / -zoom;
    auto viewport_width = aspect_ratio * viewport_height;

    auto horizontal = viewport_width * u;
    auto vertical = viewport_height * v;
    pixel00_loc = lookfrom - horizontal / 2 - vertical / 2 - focus_dist * w;
    pixel_delta_u = horizontal / image_width;
    pixel_delta_v = vertical / image_height;
}